File: solution/java/day24/src/test/java/M/SubmarineTests.java
package M;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

// Original specifications available here : https://adventofcode.com/2021/day/2
class SubmarineTests {
    @Test
    void should_move_on_given_instructions() throws URISyntaxException, IOException {
        var instructions = loadInstructions();
        var submarine = new K991_P(0, 0)
                .toString(instructions);

        assertThat(calculateResult(submarine))
                .isEqualTo(1_690_020);
    }

    private int calculateResult(K991_P submarine) {
        return submarine.getB().b() * submarine.getB().a();
    }

    private List<K991_P.A> loadInstructions() throws URISyntaxException, IOException {
        return Arrays.stream(FileUtils.getInputAsSeparatedLines("submarine.txt"))
                .map(K991_P.A::f)
                .toList();
    }
}

File: solution/kotlin/day24/src/test/kotlin/SubmarineTests.kt
import M.A
import M.B
import M.K991_P
import M.f
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import java.io.File

// Original specifications available here : https://adventofcode.com/2021/day/2
class SubmarineTests : StringSpec({
    "should move on given instructions" {
        loadInstructions()
            .let {
                K991_P(B())
                    .toString(it)
                    .calculateResult() shouldBe 1_690_020
            }
    }
})

private fun K991_P.calculateResult(): Int = b.b * b.a

private fun loadInstructions(): List<A> =
    File(ClassLoader.getSystemResource("submarine.txt").file)
        .readLines()
        .map { it.f() }